libraries/fatfs/adi_diskio.c

```diff
@@ -0,0 +1,243 @@
+		return SD_disk_status();;
+	case DEV_RAM :
+		return STA_NODISK;
+	case DEV_USB :
+		return STA_NODISK;
+	default:
+		return STA_NODISK;
+	}
+	return STA_NOINIT;
+}
+
+
+
+/*-----------------------------------------------------------------------*/
+/* Initialize a Drive                                                    */
+/*-----------------------------------------------------------------------*/
+
+DSTATUS disk_initialize (
+	BYTE pdrv				/* Physical drive nmuber to identify the drive */
+)
+{
+	switch (pdrv) {
+	case DEV_SD :
+		return SD_disk_initialize();
+	case DEV_RAM :
+		return STA_NODISK;
+	case DEV_USB :
+		return STA_NODISK;
+	}
+	return STA_NOINIT;
+}
+
+
+
+/*-----------------------------------------------------------------------*/
+/* Read Sector(s)                                                        */
+/*-----------------------------------------------------------------------*/
+
+DRESULT disk_read (
+	BYTE pdrv,		/* Physical drive nmuber to identify the drive */
+	BYTE *buff,		/* Data buffer to store read data */
+	LBA_t sector,		/* Start sector in LBA */
+	UINT count		/* Number of sectors to read */
+)
+{
+
+	switch (pdrv) {
+	case DEV_SD :
+		return SD_disk_read(buff, sector, count);
+	case DEV_RAM :
+		return RES_NOTRDY;
+	case DEV_USB :
+		return RES_NOTRDY;
+	}
+	return RES_PARERR;
+}
+
+
+
+/*-----------------------------------------------------------------------*/
+/* Write Sector(s)                                                       */
+/*-----------------------------------------------------------------------*/
+
+#if FF_FS_READONLY == 0
+
+DRESULT disk_write (
+	BYTE pdrv,			/* Physical drive nmuber to identify the drive */
+	const BYTE *buff,	/* Data to be written */
+	LBA_t sector,		/* Start sector in LBA */
+	UINT count			/* Number of sectors to write */
+)
+{
+	switch (pdrv) {
+	case DEV_SD:
+		return SD_disk_write(buff, sector, count);
+	case DEV_RAM :
+		return RES_NOTRDY;
+	case DEV_USB :
+		return RES_NOTRDY;
+	}
+
+	return RES_PARERR;
+}
+
+#endif
+
+
+/*-----------------------------------------------------------------------*/
+/* Miscellaneous Functions                                               */
+/*-----------------------------------------------------------------------*/
+
+DRESULT disk_ioctl (
+	BYTE pdrv,		/* Physical drive nmuber (0..) */
+	BYTE cmd,		/* Control code */
+	void *buff		/* Buffer to send/receive control data */
+)
+{
+	switch(pdrv) {
+	case DEV_SD:
+		switch (cmd){
+		case CTRL_SYNC: return RES_OK;
+		case GET_SECTOR_COUNT:
+			*(LBA_t *)buff = sd_desc->memory_size / DATA_BLOCK_LEN;
+			return RES_OK;
+		case GET_SECTOR_SIZE:
+			*(WORD *)buff = DATA_BLOCK_LEN;
+			return RES_OK;
+		case GET_BLOCK_SIZE:
+			*(DWORD *)buff = 1u;
+			return RES_OK;
+		default: return RES_OK;
+		}
+		return RES_PARERR;
+	case DEV_RAM:
+		return RES_NOTRDY;
+	case DEV_USB:
+		return RES_NOTRDY;
+	}
+	return RES_PARERR;
+}
+
+static int32_t init_power()
+{
+	if (adi_pwr_Init() != ADI_PWR_SUCCESS)
```

Comment:
I would make this layer generic too. I guess adi_pwr_Init() is an ADuCM specific function.